File: examples/nodejs/distributed_orchestrator.ts
import { TsgoDistributedOrchestrator, TsgoVirtualDocument } from "@corsa-bind/node";

import { isMain } from "../shared.ts";

export function runDistributedOrchestratorExample() {
  const cluster = new TsgoDistributedOrchestrator(["n1", "n2", "n3"]);
  const term = cluster.campaign("n1");
  const document = TsgoVirtualDocument.inMemory(
    "cluster",
    "/main.ts",
    "typescript",
    "let value = 1;",
  );

  cluster.openVirtualDocument(document.state());
  cluster.changeVirtualDocument(document.uri, [
    {
      range: {
        start: { line: 0, character: 12 },
        end: { line: 0, character: 13 },
      },
      text: "2",
    },
  ]);

  return {
    leaderId: cluster.leaderId(),
    node2Document: cluster.document("n2", document.uri),
    term,
  };
}

if (isMain(import.meta.url)) {
  console.log(JSON.stringify(runDistributedOrchestratorExample(), null, 2));
}

File: examples/rust/distributed_orchestrator.rs
mod support;

use std::time::Duration;

use corsa_bind_rs::{
    api::{ApiMode, ApiProfile},
    lsp::{VirtualChange, VirtualDocument},
    orchestrator::DistributedApiOrchestrator,
    runtime::block_on,
};
use serde_json::{Value, json};

fn main() -> Result<(), corsa_bind_rs::TsgoError> {
    let result = block_on(async {
        let orchestrator = DistributedApiOrchestrator::new(["n1", "n2", "n3"]);
        let profile = ApiProfile::new(
            "distributed-demo",
            support::mock_api_config("distributed_orchestrator", ApiMode::AsyncJsonRpcStdio)?,
        );
        let term = orchestrator.campaign("n1")?;
        let document =
            VirtualDocument::in_memory("cluster", "/main.ts", "typescript", "let value = 1;")?;
        orchestrator.open_virtual_document("n1", document.clone())?;
        let updated = orchestrator.change_virtual_document(
            "n1",
            &document.uri,
            [VirtualChange::splice(
                lsp_types::Range::new(
                    lsp_types::Position::new(0, 12),
                    lsp_types::Position::new(0, 13),
                ),
                "2",
            )],
        )?;
        let cached_ping: Value = orchestrator
            .cached(
                &profile,
                "n1",
                "ping",
                Some(Duration::from_secs(30)),
                |client| async move { client.raw_json_request("ping", Value::Null).await },
            )
            .await?;
        let follower_document = orchestrator.document("n2", &document.uri).ok_or_else(|| {
            corsa_bind_rs::TsgoError::Protocol(
                "distributed orchestrator example did not replicate to follower".into(),
            )
        })?;

        Ok::<_, corsa_bind_rs::TsgoError>(json!({
            "leaderId": orchestrator.leader_id(),
            "term": term,
            "cachedPing": cached_ping,
            "updatedDocument": updated,
            "followerDocument": follower_document,
        }))
    })?;

    support::print_json(result);
    Ok(())
}
